WordPressKit/WordPressKit/MediaServiceRemoteXMLRPC.m

```diff
@@ -30,13 +30,40 @@ - (void)getMediaWithID:(NSNumber *)mediaID
 - (void)getMediaLibraryWithSuccess:(void (^)(NSArray *))success
                            failure:(void (^)(NSError *))failure
 {
-    NSArray *parameters = [self defaultXMLRPCArguments];
+    NSMutableArray * mediaResult = [[NSMutableArray alloc] init];
+    [self getMediaLibraryStartOffset:0 media:mediaResult success:success failure:failure];
+}
+
+- (void)getMediaLibraryStartOffset:(NSUInteger)offset
+                             media:(NSMutableArray *)media
+                           success:(void (^)(NSArray *))success
+                           failure:(void (^)(NSError *))failure
+{
+    NSInteger pageSize = 100;
+    NSDictionary *filter = @{
+                                @"number": @(pageSize),
+                                @"offset": @(offset)
+                            };
+    NSArray *parameters = [self XMLRPCArgumentsWithExtra:filter];
+
     [self.api callMethod:@"wp.getMediaLibrary"
               parameters:parameters
                  success:^(id responseObject, NSHTTPURLResponse *httpResponse) {
                      NSAssert([responseObject isKindOfClass:[NSArray class]], @"Response should be an array.");
                      if (success) {
-                         success([self remoteMediaFromXMLRPCArray:responseObject]);
+                         NSArray *pageMedia = [self remoteMediaFromXMLRPCArray:responseObject];
+                         if (pageMedia.count > 0) {
+                             [media addObjectsFromArray: pageMedia];
+                         }
+                         // Did we got all the items we requested or it's finished?
+                         if (pageMedia.count < pageSize) {
+                             if (success) {
+                                 success(media);
+                             }
+                             return;
+                         }
+                         NSUInteger newOffset = offset + pageMedia.count;
+                         [self getMediaLibraryStartOffset:newOffset media:media success: success failure: failure];
                      }
                  }
                  failure:^(NSError *error, NSHTTPURLResponse *httpResponse) {
```

Comment:
Should we fail entirely if we managed to receive some of the data? If we received 150 / 250 items, I think it would be best to display those, even if we also display an error?

Reply:
One of the problems here is that when we merge the data we delete anything that is not part of the result. If we are happy with that we could respond success with partial data.

Reply:
That makes sense. We could potentially delete media that loaded successfully once but then failed the next time the library was refreshed. Seems safer just to keep it?

The alternative might be to check the local media library count for this site, and if the count is 0 or less than the amount we fetched, we'll keep whatever we managed to fetch?